src/inpod/workloadmanager.rs

```diff
@@ -0,0 +1,606 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::readiness;
+use drain::Watch;
+use std::path::PathBuf;
+use std::time::Duration;
+use tokio::net::UnixStream;
+use tracing::{debug, error, info, warn};
+
+use super::statemanager::WorkloadProxyManagerState;
+use super::Error;
+
+#[mockall_double::double]
+use super::protocol::WorkloadStreamProcessor;
+
+const RETRY_DURATION: Duration = Duration::from_secs(5);
+
+struct WorkloadProxyNetworkHandler {
+    uds: PathBuf,
+}
+
+struct WorkloadProxyReadinessHandler {
+    ready: readiness::Ready,
+    // Manually drop as we don't want to mark ready if we are dropped.
+    // This can happen when the server drains.
+    block_ready: Option<std::mem::ManuallyDrop<readiness::BlockReady>>,
+}
+
+pub struct WorkloadProxyManager {
+    state: super::statemanager::WorkloadProxyManagerState,
+    networking: WorkloadProxyNetworkHandler,
+    // readiness - we are only ready when we are connected. if we get disconnected, we become not ready.
+    readiness: WorkloadProxyReadinessHandler,
+}
+
+struct WorkloadProxyManagerProcessor<'a> {
+    state: &'a mut super::statemanager::WorkloadProxyManagerState,
+    readiness: &'a mut WorkloadProxyReadinessHandler,
+
+    next_pending_retry: Option<std::pin::Pin<Box<tokio::time::Sleep>>>,
+}
+
+impl WorkloadProxyReadinessHandler {
+    fn new(ready: readiness::Ready) -> Self {
+        let mut r = Self {
+            ready,
+            block_ready: None,
+        };
+        r.not_ready();
+        r
+    }
+    fn mark_ready(&mut self) {
+        // take the block ready out of the ManuallyDrop, and drop it.
+        if self.block_ready.is_some() {
+            debug!("workload proxy manager is ready");
+            let block_ready: Option<readiness::BlockReady> = self
+                .block_ready
+                .take()
+                .map(std::mem::ManuallyDrop::into_inner);
+
+            std::mem::drop(block_ready);
+        }
+    }
+
+    fn not_ready(&mut self) {
```

Comment:
we dont "need" this per-say - as you said there's no service backing the ztunnel. But by the same logic, why have any readiness signal in the ztunnel at all?

It seems that in ztunnel the semantics of "readiness" is that ztunnel is ready to handle pods. It seems that the motivation here is to signal to the user that something requires immediate attention.

Using that semantics, if the ztunnel is not connected to the node agent, it is not "ready".
While envoy won't stop being ready, it does have a metric to indicate something similar (connection to the control-plane, called "connected_state").

If you think this will confuse users rather than help, I can remove this.

Reply:
Readiness does still have some meaning at startup even without a service, just not _after_ startup; you need to become ready before other instances will terminate in a rollout.

Reply:
Note that this doesn't currently apply in the ztunnel case - AFAIK we cannot run ztunnel side by side with another ztunnel.. so currently other instance will have to terminate first.

Even if we had ztunnel working SxS, the current logic is consistent with the rollout semantics as well - You wouldn't want to remove an old ztunnel if the new one is not connected to the node agent.

I think this mainly comes to istio users expectation. If we think that readiness changing in back to "not ready"  will create confusion, I can replace this with a metric.

Reply:
> Note that this doesn't currently apply in the ztunnel case - AFAIK we cannot run ztunnel side by side with another ztunnel.. so currently other instance will have to terminate first.
> 
> Even if we had ztunnel working SxS, the current logic is consistent with the rollout semantics as well - You wouldn't want to remove an old ztunnel if the new one is not connected to the node agent.

Right - if the new ztunnel is not "ready" because it cannot talk to the current CNI agent on the node, and the old one is/can, the rollout *should* stall, otherwise users end up in a bad state with no explicit signaling.

Again though, this only applies to tirekicking. Rollout replace of ztunnel isn't recommended for production upgrades.

Reply:
I 100% agree it needs to sync to the agent at startup. I just don't see why it should become unready if it _later_ disconnects. Note a disconnect happens in regular operation, too.

Reply:
Oh I see what you're saying - yeah that's fair. I can't think of a situation during rollout where:

1. ztunnel A would have previously been able to connect to the CNI during startup.
2. ztunnel A ceases to be able to for `$SOME_REASON`.
3. A rollout happens, and the new ztunnel B _IS_ able to connect to the same CNI during startup.